### DISCUSSION TITLE
Bootstrapping Starlark types

### DISCUSSION THREAD

**Discussion by comius** [bazelbuild/bazel#23168] [Design proposals]
https://docs.google.com/document/d/1Sid7EAbBd_w_T7D94Li_f_bK3zMTztFbzIMvcpzo1wY/edit?usp=sharing

**Comment by albertocavalcante** [comment]
Thanks for that!

Related Links:

- https://github.com/bazelbuild/bazel/issues/24407
- https://github.com/bazelbuild/bazel/issues/22935
- https://github.com/bazelbuild/bazel/issues/7468
- https://github.com/bazelbuild/starlark/issues/106
- https://github.com/bazelbuild/starlark/pull/303
- https://github.com/bazelbuild/bazel/wiki/Project-Ideas#static-type-checker-for-starlark
- https://github.com/withered-magic/starpls?tab=readme-ov-file#tips-and-tricks
- https://groups.google.com/g/bazel-dev/c/u-ass4pbK7U/m/aGQF72cjCAAJ
- https://groups.google.com/g/bazel-dev/c/Pk9VrPCqby0/m/zoOL1IkxEAAJ

**Comment by brandjon** [comment]
See [Starlark type system design](https://docs.google.com/document/d/17bGpij0gPWJFbxHLKTTwfKj8dAldD1kr8MYAa0zMUMs/edit?usp=sharing) for a bit more discussion on the current design and features we expect to support. Doesn't quite supersede the old doc, which still has a lot of good discussion and some unresolved questions.

**Reply by albertocavalcante** [reply]
@brandjon thanks for sharing this!! I've also seen the PR to https://github.com/bazelbuild/proposals/pull/413/files.

So far, it seems the focus has been on the Bazel domain for `.bzl` files, with Starlark-Java currently implementing this feature.

A question: I would love to see this being adopted in `starlark-go`, and I wonder if this opens the road to that. 

I've seen the current `starlark-with-types`. Currently lives under a branch, as documented at https://bazel.build/rules/language#StarlarkTypes, and, per @adonovan at https://github.com/google/starlark-go/issues/610#issuecomment-3275143082, this will depend on the spec being changed first.

`starlark-rust` just called it out experimental and implemented it https://github.com/facebook/starlark-rust/blob/main/docs/types.md. 

So my goal here would be to understand whether and when it would be possible to get a green light to start contributing type annotations to `starlark-go`.

Thanks!

(Maybe this conversation is more appropriate for https://github.com/bazelbuild/starlark/issues/106, as this discussion is under Bazel, I don't know.)
